### PR TITLE
mkdir: do not return errors for incorrect directory modes

### PR DESCRIPTION
**Pull request by cyphar**
We've had several examples of unexpected semantics with how modes are
calculated, and there will likely be many more in the future. To avoid
further regressions, we should just downgrade the error to a warning.

However, we can keep the owner checks (those -- hopefully -- shouldn't
have too many weird semantics).

See https://github.com/opencontainers/runc/issues/4408
Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,19 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased] ##
 
+### Fixed ###
+- The mode verification logic in `MkdirAll` has been adjusted to avoid false
+  positives. Now, `MkdirAll` will no longer return an error if the mode of a
+  directory we created is different to what we expected. At most you will get a
+  warning (using `github.com/sirupsen/logrus`) with information
```
Comment (kolyshkin) [inline CHANGELOG.md:13]: Either an unfinished sentence, or a missing period.

File: mkdir_linux.go [hunk 2]
```diff
@@ -185,8 +188,16 @@ func MkdirAllHandle(root *os.File, unsafePath string, mode int) (_ *os.File, Err
 		if stat, err := fstat(currentDir); err != nil {
 			return nil, fmt.Errorf("check newly created directory: %w", err)
 		} else {
-			if stat.Mode != expectedMode {
-				return nil, fmt.Errorf("%w: newly created directory %q has incorrect mode 0o%.3o (expected 0o%.3o)", errPossibleAttack, currentDir.Name(), stat.Mode, expectedMode)
+			if stat.Mode&^maxExpectedMode != 0 {
+				// Unfortunately, we can't return an error. Filesystems like
+				// vfat don't support Unix-like DAC modes, and we have had
+				// spurrious errors from this check a few times so we should
```
Comment (kolyshkin) [inline mkdir_linux.go:194]: A typo? (spurious)

**Review by kolyshkin — COMMENTED** [review]
As for vendoring, it's not really needed. Modern go can do fine without (it will download the missing modules from the proxy.golang.org to $GOMODCACHE, when needed, if they are not yet there).

Having `vendor` has its pros and contras; I haven't yet grasped which way is better, but it is definitely not a must have.

**Comment by cyphar** [comment]
Yeah, I might just drop it for now. We need `vendor` for distro packages but `filepath-securejoin` is not packaged by itself so it isn't necessary. I was just a bit surprised I forgot to include it before.

File: mkdir_linux.go [hunk 3]
```diff
@@ -185,8 +188,16 @@ func MkdirAllHandle(root *os.File, unsafePath string, mode int) (_ *os.File, Err
...
+			if stat.Mode&^maxExpectedMode != 0 {
+				// Unfortunately, we can't return an error. Filesystems like
+				// vfat don't support Unix-like DAC modes, and we have had
+				// spurious errors from this check a few times so we should
+				// play it safe to avoid regressions. The uid/gid check is more
+				// important and to be honest these checks are not that
+				// important (they mostly exist out of an abundance of caution
+				// -- it's not entirely clear what practical attack this is
+				// defending against).
+				logrus.Printf("filepath-securejoin: %v: newly created directory %q has incorrect mode 0o%.3o (expected 0o%.3o)", errPossibleAttack, currentDir.Name(), stat.Mode, maxExpectedMode)
```
Comment (kolyshkin) [inline mkdir_linux.go:200]: nit: perhaps it's better to use `logrus.Infof` (to be explicit about the log level -- otherwise Printf is the same as Infof).

**Review by kolyshkin — APPROVED** [review]
lgtm except one nit

File: mkdir_linux.go (hunk 3 above)
Comment (cyphar) [reply mkdir_linux.go:200]: Ah, this is from when I used the stdlib `log`. Fixed.

**Comment by cyphar** [comment]
Should it be a warning or info @kolyshkin?

**Comment by kolyshkin** [comment]
> Should it be a warning or info @kolyshkin?

Either way is fine with me. Info is one level below Warning (and just above Debug), and the default level is Warning.

Meaning, if you want it to show by default, use `Warnf`. If not -- `Infof` or even `Debugf`.

That's from the top of my head and need to be checked.

**Comment by cyphar** [comment]
Default logrus includes output for `info` but runc sets the default to `warn` iirc.

FWIW, after speaking to @rata and @alban I am convinced these checks don't defend against any practical attack so it might make sense to just drop these in the future. The owner and mode checks are also going to be wrong for `vfat` if you mount it with `uid=1234,gid=5678,umask=0`.

**Comment by cyphar** [comment]
I ended up dropping the warnings since it's not clear anyone would care anyway, and tools that don't use logrus would get warnings they can't practically block.

**Pull request merged**